aten/src/ATen/native/cuda/Shape.cu

```diff
@@ -0,0 +1,399 @@
+  // wasn't possible to cat empty tensors unless all the other tensors were
+  // 1-dimensional, so we allowed these tensors to be "skipped".  We maintain
+  // this behavior for backwards compatibility, but only for this specific size
+  // (i.e. other empty sizes are not skipped).
+  // FIXME: warn if this is the case
+  auto should_skip = [](const Tensor &t) {
+    return t.dim() == 1 && at::native::size(t, 0) == 0;
+  };
+  bool hasSkippedInput = false;
+
+  const Tensor *notSkippedTensor = NULL;  // non-owning reference
+  int nDims = 0;
+
+  // Inputs cannot alias the output tensor
+  for (int i = 0; i < inputs.size(); i++) {
+    auto lap = at::get_overlap_status(out, inputs[i]);
+    TORCH_CHECK(lap != at::MemOverlapStatus::PARTIAL &&
+                lap != at::MemOverlapStatus::FULL,
+                "unsupported operation: the input tensors cannot refer to any "
+                "of the output memory locations. Found overlap in input "
+                "tensor ", i);
+  }
+
+  for (int i = 0; i < inputs.size(); i++)
+  {
+    if (should_skip(inputs[i])) {
+      hasSkippedInput = true;
+      continue;
+    }
+    nDims = inputs[i].dim();
+    notSkippedTensor = &inputs[i];
+  }
+
+  // If all inputs are empty tensors, return an empty tensor
+  if (notSkippedTensor == NULL) {
+    return out;
+  }
+
+  TORCH_CHECK(inputs.size() > 0, "invalid number of inputs ", inputs.size());
+  TORCH_CHECK(dimension >= 0, "invalid dimension ", dimension);
+
+  std::vector<int64_t> size(nDims);
+
+  // Compute size of the result in the cat dimension
+  int64_t cat_dim_size = 0;
+  for (int i = 0; i < inputs.size(); i++) {
+    const Tensor &tensor = inputs[i];
+    if (should_skip(tensor)) {
+      continue;
+    }
+    check_shape_except_dim(*notSkippedTensor, tensor, dimension);
+    cat_dim_size += dim_size(tensor, dimension);
+  }
+
+  // Compute the size of the result
+  for (int dim = 0; dim < nDims; dim++) {
+    int64_t out_dim_size = dim_size(*notSkippedTensor, dim);
+    if (dim == dimension) {
+      out_dim_size = cat_dim_size;
+    }
+    size[dim] = out_dim_size;
+  }
+  out.resize_(size);
+  if (out.numel() == 0) {
+    return out;
+  }
+
+  // We parallelize the copy if all 6 conditions pass:
+  //
+  // 1. There is more than one input tensor
+  // 2. No empty inputs
+  // 3. The out tensor is 32-bit indexable
+  // 4. The number of dimensions is <= 4
+  // 5. All input tensors are contiguous (output tensor may be non-contig)
+  // 6. All input tensors can use 32-bit indexing
+  // 7. All input tensors are on the same device
+
+  const bool all32BitIndexable = std::all_of(inputs.begin(), inputs.end(),
+    [] (const Tensor& t) {
+      // Special case for an empty tensor is required to remain compatible with
+      // the legacy behaviour of THCTensor_canUse32BitIndexMath which skips
+      // all computation for empty tensors (because it employs
+      // THCTensor_nDimensionLegacyAll)
+      return t.numel() == 0
```

Comment:
swap the conditionals here so that more common is evaluated first.

Reply:
If the first one is false, the second one will crash.

Reply:
if `canUse32BitIndexMath` crashes with empty tensor, it makes sense to fix it, rather than workaround here. The comment above is misleading and should be deleted btw, you are not calling THCTensor_canUse32BitIndexMath.